src/rgw/rgw_op.cc

```diff
@@ -1295,12 +1299,14 @@ void RGWGetObj::execute()
   gc_invalidate_time = ceph_clock_now(s->cct);
   gc_invalidate_time += (s->cct->_conf->rgw_gc_obj_min_wait / 2);
 
-  RGWGetObj_CB cb(this);
-
+  RGWGetObj_CB* cb = new RGWGetObj_CB(this);
+  RGWGetDataCB* filter = (RGWGetDataCB*)cb;
+  RGWGetDataCB* decrypt = nullptr;
```

Comment:
i'd like to find a way to use these abstractions without the heap allocations. we should be able to keep `cb` on the stack as it was, as long as `~RGWPutObj_Filter()` doesn't try to delete it. the Decompress filter is tricky because it's constructed later on only when `need_decompress`, but `boost::optional` seems like a good fit. for example:
```c++
  RGWGetObj_CB cb(this);
  RGWGetDataCB* filter = &cb;
  boost::optional<RGWGetObj_Decompress> decompress;
```
..and later:
```c++
  if (need_decompress) {
    ...
    decompress = boost::in_place(s->cct, cs_info, partial_content, filter);
    filter = &*decompress;
  }
```
what do you think?